src/StructureMap.DynamicInterception/MethodInvocation.cs

```diff
@@ -10,28 +10,43 @@ namespace StructureMap.DynamicInterception
     internal class MethodInvocation : ISyncMethodInvocation, IAsyncMethodInvocation
     {
         private readonly IInvocation _invocation;
-        private readonly IList<IArgument> _arguments;
-        private readonly IDictionary<string, IArgument> _argumentMap;
+        private readonly Lazy<List<IArgument>> _arguments;
+        private readonly Lazy<IDictionary<string, IArgument>> _argumentMap;
 
         public MethodInvocation(IInvocation invocation)
         {
             _invocation = invocation;
-            _arguments = invocation.Method.GetParameters()
-                .Zip(invocation.Arguments, (info, value) => new { info, value })
-                .Select((t, i) => new Argument(invocation, i, t.value, t.info))
-                .ToList<IArgument>();
-            _argumentMap = _arguments.ToDictionary(a => a.ParameterInfo.Name);
+            _arguments = new Lazy<List<IArgument>>(() => invocation.Method.GetParameters()
+                                                                    .Zip(invocation.Arguments, (info, value) => new { info, value })
+                                                                    .Select((t, i) => new Argument(invocation, i, t.value, t.info))
+                                                                    .ToList<IArgument>());
+            _argumentMap = new Lazy<IDictionary<string, IArgument>>(() => _arguments.Value.ToDictionary(a => a.ParameterInfo.Name));
         }
 
         public IList<IArgument> Arguments
         {
-            get { return new List<IArgument>(_arguments); }
+            get { return _arguments.Value.AsReadOnly(); }
```

Comment:
`AsReadOnly()` is documented as an O(1) operation, so not much performance savings. :) But it does let us change the type of the variable to a `ReadOnlyCollection` to ensure that the collection is not modified after being created.